PayButton/paybutton-server
# [#660] feat: big address sync improvement

Related to #660


Depends on
---
- [x] #724
- [x] #725 
- [x] #726 
- [x] #727 




Description
---
Improves the performance when inserting transactions.

This is achieved using for-loops instead of async concurrent promises to insert the txs on the database, while running every batch of the tx insertions inside a DB transaction.



The plots below show how this considerably brings down the CPU usage without diminishing the insertion speed too much. For all of them, the beginning should be seen as the initial sync of addresses + spinning up of the app, then a hiatus of nothing being synced and then, finally,  the addition of a large address (in all examples [qrcuqadqrzp2uztjl9wn5sthepkg22majylrprqkuk](https://ecashexplorer.com/address/ecash:qrcuqadqrzp2uztjl9wn5sthepkg22majylrprqkuk) was used). This last part of the plot is where the tx sync of a large address is happening.

## Master:
![performance_plot](https://github.com/PayButton/paybutton-server/assets/21281174/3ce09598-4ad1-4049-97de-d00eb8d1a4cb)



## This branch:
![performance_plot](https://github.com/PayButton/paybutton-server/assets/21281174/e7656344-7dfb-4cd6-bb9f-ebceb08dd064)


Test plan
---


1. Make sure you have python and the python packages `matplotlib` and `dotenv` installed.
2. Spin up the containers and then run `python scripts/performance.py`. On the plots above, I've used `make reset-dev && sleep 20; python scripts/performance.py `
3. Log in the dev user and wait until the activity in the plot is close to 0
4. Add a button with a large address.
5. Check the performance.
6. To stop running the performance script, close the plot and immediately press CTRL+C.
7.  A `plot_performance.jpg` file will be saved in the repository root dir.
